zebra-state/src/service/check/tests/prop.rs

```diff
@@ -358,32 +888,52 @@ fn new_state_with_mainnet_genesis() -> (StateService, FinalizedBlock) {
 }
 
 /// Make sure the supplied nullifiers are distinct, modifying them if necessary.
-fn make_distinct_nullifiers<'joinsplit>(
-    nullifiers: impl IntoIterator<Item = &'joinsplit mut sprout::Nullifier>,
-) {
+fn make_distinct_nullifiers<'until_modified, NullifierT>(
+    nullifiers: impl IntoIterator<Item = &'until_modified mut NullifierT>,
+) where
+    NullifierT: Into<[u8; 32]> + Clone + Eq + std::hash::Hash + 'until_modified,
+    [u8; 32]: Into<NullifierT>,
+{
     let nullifiers: Vec<_> = nullifiers.into_iter().collect();
 
     if nullifiers.iter().unique().count() < nullifiers.len() {
         let mut tweak: u8 = 0x00;
         for nullifier in nullifiers {
-            nullifier.0[0] = tweak;
+            let mut nullifier_bytes: [u8; 32] = nullifier.clone().into();
+            nullifier_bytes[0] = tweak;
+            *nullifier = nullifier_bytes.into();
+
             tweak = tweak
                 .checked_add(0x01)
                 .expect("unexpectedly large nullifier list");
         }
     }
 }
 
-/// Return a `Transaction::V4` containing `joinsplit_data`.
+/// Return a `Transaction::V4` containing `joinsplit_data`,
+/// with its `JoinSplit`s replaced by `joinsplits`.
 ///
 /// Other fields have empty or default values.
-fn transaction_v4_with_joinsplit_data(
+///
+/// # Panics
+///
+/// If there are no `JoinSplit`s in `joinsplits`.
+fn transaction_v4_with_joinsplit_data<'until_cloned>(
     joinsplit_data: impl Into<Option<JoinSplitData<Groth16Proof>>>,
+    joinsplits: impl IntoIterator<Item = &'until_cloned JoinSplit<Groth16Proof>>,
 ) -> Transaction {
     let mut joinsplit_data = joinsplit_data.into();
+    let joinsplits: Vec<_> = joinsplits.into_iter().cloned().collect();
```

Comment:
Suggestion: an idea to try to simplify the code might be to receive owned items. But it does require updating the calls to remove the `&`s.

```suggestion
fn transaction_v4_with_joinsplit_data(
    joinsplit_data: impl Into<Option<JoinSplitData<Groth16Proof>>>,
    joinsplits: impl IntoIterator<Item = JoinSplit<Groth16Proof>>,
) -> Transaction {
    let mut joinsplit_data = joinsplit_data.into();
    let joinsplits: Vec<_> = joinsplits.into_iter().collect();
```

Reply:
Fixed in e6ba566d by passing vectors rather than slices

Reply:
Great! Just a FYI, it doesn't need the `vec!` part. Just `[a, b]` should work as well since Rust 1.53 adds `impl IntoIterator` for arrays.

Reply:
Ah right yeah type inference didn't work with slices because of the temporary exceptions to that feature. But now we're taking an owned iterator it works fine.

See 239de32a